opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/buckets.py

```diff
@@ -0,0 +1,148 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+class Buckets:
+    def __init__(self):
+        self._counts = [0]
+
+        # The term "index" refers to the number of the
+        # histogram bucket used to determine its boundaries.
+        # The lower-boundary of a bucket is determined by
+        # formula base**index and the upper-boundary of a
+        # bucket is base ** (index + 1). Index values are signed
+        # to account for values less than or equal to 1.
+
+        # self._index_* will all have values equal to a certain index that is
+        # determined by the corresponding mapping _map_to_index function and
+        # the value of the index depends on the value passed to _map_to_index.
+
+        # Index of the 0th position in the backing array: backing[0] is the
+        # count in the bucket with index self._index_base.
+        self._index_base = 0
+
+        # indexStart is the smallest index value represented in the backing
+        # array.
+        self._index_start = 0
+
+        # indexEnd is the largest index value represented in the backing array.
+        self._index_end = 0
+
+    @property
+    def counts(self):
+        return self._counts
+
+    def grow(self, needed: int, max_size: int):
+
+        size = len(self._counts)
+        bias = self._index_base - self._index_start
+        old_positive_limit = size - bias
+
+        needed = needed - 1
+
+        needed |= needed >> 1
+        needed |= needed >> 2
+        needed |= needed >> 4
+        needed |= needed >> 8
+        needed |= needed >> 16
```

Comment:
I can't remember what this does? 🤣  (j/k but please add a comment)

Reply:
Right, reimplemented and added a comment.